mlflow/entities/span.py

```diff
@@ -14,3 +40,434 @@ class SpanType:
+        """
+        return self._attributes.get_all()
+
+    @property
+    def events(self) -> List[SpanEvent]:
+        """
+        Get all events of the span.
+
+        Returns:
+            A list of all events of the span.
+        """
+        return [
+            SpanEvent(
+                name=event.name,
+                timestamp=event.timestamp,
+                # Convert from OpenTelemetry's BoundedAttributes class to a simple dict
+                # to avoid the serialization issue due to having a lock object.
+                attributes=dict(event.attributes),
+            )
+            for event in self._span.events
+        ]
+
+    def get_attribute(self, key: str) -> Optional[Any]:
+        """
+        Get a single attribute value from the span.
+
+        Args:
+            key: The key of the attribute to get.
+
+        Returns:
+            The value of the attribute if it exists, otherwise None.
+        """
+        return self._attributes.get(key)
+
+    def to_dict(self):
+        # NB: OpenTelemetry Span has to_json() method, but it will write many fields that
+        #  we don't use e.g. links, kind, resource, trace_state, etc. So we manually
+        #  cherry-pick the fields we need here.
+        return {
+            "name": self.name,
+            "context": {
+                "span_id": self.span_id,
+                "trace_id": self._trace_id,
+            },
+            "parent_id": self.parent_id,
+            "start_time": self.start_time_ns,
+            "end_time": self.end_time_ns,
+            "status_code": self.status.status_code,
+            "status_message": self.status.description,
+            "attributes": dict(self._span.attributes),
+            "events": [asdict(event) for event in self.events],
+        }
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "Span":
+        """
+        Create a Span object from the given dictionary.
+        """
+        try:
+            request_id = data.get("attributes", {}).get(SpanAttributeKey.REQUEST_ID)
+            if not request_id:
+                raise MlflowException(
+                    f"The {SpanAttributeKey.REQUEST_ID} attribute is empty or missing.",
+                    INVALID_PARAMETER_VALUE,
+                )
+
+            trace_id = decode_id(data["context"]["trace_id"])
+            span_id = decode_id(data["context"]["span_id"])
+            parent_id = decode_id(data["parent_id"]) if data["parent_id"] else None
+
+            otel_span = OTelReadableSpan(
+                name=data["name"],
+                context=build_otel_context(trace_id, span_id),
+                parent=build_otel_context(trace_id, parent_id) if parent_id else None,
+                start_time=data["start_time"],
+                end_time=data["end_time"],
+                attributes=data["attributes"],
+                status=SpanStatus(data["status_code"], data["status_message"]).to_otel_status(),
```

Comment:
Can we directly use OTelStatus(data["status_code"], data["status_message"])?

Reply:
OTel's Status only accepts their enum ([ref](https://github.com/mlflow/mlflow/blob/tracing/mlflow/entities/span_status.py#L59)) so using our constructor is easier indeed🙂